### PR TITLE
IPFS resolver fixes and parallelization

### PR DESCRIPTION
**Pull request by nerfZael**

### CODE REVIEW

File: packages/js/plugins/ipfs/src/exec-ipfs/CancelablePromise.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+export type CancelablePromise<TReturn> = {
```
Comment (Niraj-Kamdar) [inline packages/js/plugins/ipfs/src/exec-ipfs/CancelablePromise.ts]: Move this to types

File: packages/js/plugins/ipfs/src/exec-ipfs/cancelableExecIpfs.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    provider: string,
+    options: unknown
+  ) => Promise<TReturn>,
+): CancelablePromise<TReturn> => {
+  let controller = new AbortController();
+  
+  let error: any = undefined;
+  
+  //If timer is not 0 then set a timeout to abort the execution
+  const timer: NodeJS.Timeout | undefined = timeout
```
Comment (Niraj-Kamdar) [inline packages/js/plugins/ipfs/src/exec-ipfs/cancelableExecIpfs.ts]: Will this type anotation work on browser environment?

Comment (Niraj-Kamdar) [reply packages/js/plugins/ipfs/src/exec-ipfs/cancelableExecIpfs.ts]: I think letting it infer type would be better

File: packages/js/plugins/ipfs/src/exec-ipfs/execIpfsWithProviders.ts [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    cancelAllRequests(requests);
+  } else {
+    //Throw all aggregated errors
+    throw new Error(errors.map(x => x.message).join("\n"));
+  }
+
+  return response.result as TReturn;
+};
+
+const gatherSuccessPromises = async <TReturn>(
```
Comment (Niraj-Kamdar) [inline packages/js/plugins/ipfs/src/exec-ipfs/execIpfsWithProviders.ts]: This is a neat trick!

**Review by Niraj-Kamdar — DISMISSED** [review]
LGTM! Really like the clever use of promises!

File: packages/js/plugins/ipfs/src/exec-ipfs/CancelablePromise.ts (hunk 1 above)
Comment (nerfZael) [reply packages/js/plugins/ipfs/src/exec-ipfs/CancelablePromise.ts]: I can add a new dir here in exec-ipfs called types, since I don't intend on exporting this out

File: packages/js/plugins/ipfs/src/exec-ipfs/cancelableExecIpfs.ts (hunk 2 above)
Comment (nerfZael) [reply packages/js/plugins/ipfs/src/exec-ipfs/cancelableExecIpfs.ts]: cant infer because clearTimeout won't work then, but I'll try and work around it

File: packages/js/plugins/ipfs/src/exec-ipfs/execIpfsWithProviders.ts (hunk 3 above)
Comment (nerfZael) [reply packages/js/plugins/ipfs/src/exec-ipfs/execIpfsWithProviders.ts]: Was fun coding it :)

File: packages/js/plugins/ipfs/schema.graphql [hunk 4]
```diff
@@ -24,10 +24,25 @@ type ResolveResult {
 }
 
 type Options {
```
Comment (Niraj-Kamdar) [inline packages/js/plugins/ipfs/schema.graphql:26]: It'd be nice if we can have inside the ENV since all the call can share this vars, We also have option to override env at query time so I don't see any downside in using ENV here.

**Review by Niraj-Kamdar — CHANGES_REQUESTED** [review]

**Review by Niraj-Kamdar — CHANGES_REQUESTED** [review]
Check the comment

File: packages/js/plugins/ipfs/schema.graphql (hunk 4 above)
Comment (nerfZael) [reply packages/js/plugins/ipfs/schema.graphql:26]: I'd agree if we had the IPFS resolver and IPFS sdk plugins separated, but right now it gets confusing.
The Options here are used for the IPFS sdk part and is configurable per request (while using a single provider) while the QueryEnv down below is used for IPFS resolver part.

When we separate them we can have Env for both and configure them without the overlapping mess. 
TLDR: I'd hold off on this until we separate the plugins and leave this as is right now
Thoughts?

Comment (Niraj-Kamdar) [reply packages/js/plugins/ipfs/schema.graphql:26]: That makes sense! And TBH it will only be this single release for which we will have this. In next release we need to refactor plugins anyway!

**Review by dOrgJelli — APPROVED** [review]
